### PR TITLE
[B] Effect, xp and clear commands can target incorrect players - Fixes BUKKIT-5356

### PR DESCRIPTION
**Pull request by stuntguy3000**
**The Issue:**
For 3 commands (effect, xp and clear) the player entered may not always be the player that receives the effect of the command. The code is told to find a player with the name, where it should find the exact player. One example is the clear command, an OP may accidentally remove a players inventory causing havoc.

Example: Two online players, "stunt" and "stuntguy3000". If i wanted to clear the inventory of stuntguy3000, if i were to type /clear stunt the inventory of "stunt" would be cleared. This is not a very big issue, however I have ran into the issue before. Once again, this also helps to mimic vanilla behavior.

**Justification for this PR:**
This PR prevents possible accidental mishaps, at the loss of being able to shortcut names (eg having to put stuntguy instead of stuntguy3000). It is crucial that the user entered is the user that receives the effect of the command, because the three commands which this effects can cause damage (mainly the clear inventory command).

This also helps to mimic vanilla behavior.

**JIRA Ticket:** https://bukkit.atlassian.net/browse/BUKKIT-5356

**Testing Material**
Compiled changes from local fork and tested.

**Associated Bukkit Ticket:** https://github.com/Bukkit/Bukkit/pull/1019

### CODE REVIEW

**Comment by turt2live** [comment]
Honestly, I prefer the shortcuts.

**Comment by Wolvereness** [comment]
Associated [ticket](https://bukkit.atlassian.net/browse/BUKKIT-5356) closed.

**Comment by stuntguy3000** [comment]
"Wesley Wolfe added a comment
No examples provided of how this is a bug. Being able to use shortened named is a feature, and when combined with features like tab-completion, I don't see any justification that this is a bug."

Example: Two online players, "stunt" and "stuntguy3000". If i wanted to clear the inventory of stuntguy3000, if i were to type /clear stunt the inventory of "stunt" would be cleared. This is not a very big issue, however I have ran into the issue before. Once again, this also helps to mimic vanilla behavior.

**Comment by turt2live** [comment]
You just described expected behavior.

sent from mobile

**Comment by stuntguy3000** [comment]
I described it wrongly. Let me rephrase. 

Example: Two online players, "stunt" and "stuntguy3000". If i wanted to
 clear the inventory of stuntguy3000, if i were to type /clear stu the
 inventory of "stunt" would be cleared. This is not a very big issue,
 however I have ran into the issue before. Once again, this also helps to
 mimic vanilla behavior.

This is a fault on the players end, yes, however this issue does not exist with /op and /pardon (being two examples) as it does not find the player with the closest name, rather the exact name. I believe that it should be implemented to function like /op an /pardon to prevent these mishaps (Also to keep consistency).

**Comment by turt2live** [comment]
Which is still expected behavior

sent from mobile

**Comment by stuntguy3000** [comment]
I know that, however I am saying it shouldn't be expected and (one example) when you are administering a server with more players than they eye can see, these mishaps could happen all the time.

There is a good reason /op op's the specified player, and not finds the closest match.

**Comment by turt2live** [comment]
As an administrator, I'm aware of the tab complete method and understand
how the arguments work by reading their documentation.

Not sure what you're trying to argue at this point.

sent from mobile

**Comment by stuntguy3000** [comment]
I have three real points

1) This helps to keep consistency with other commands such as /op and /pardon
2) Some administrators don't know of tabbing and may be ignorant of how the command properly functions (I have seen it happen, as my example shows).
3) This mimics vanilla functionality, where the exact player entered is what is used (Which is clearly important)

Points 1 and 3 are the main arguments.

**Comment by mbax** [comment]
The commands you mention are usable on offline players ;)

**Comment by turt2live** [comment]
And here are my counter arguments:

1) op and pardon are strict for an obvious reason. Clear doesn't need to be
strict
2) I don't see how Bukkit should protect a user from ignorance
3) this would have been considered when implementing the commands, and the
result of that consideration is clear

sent from mobile

**Comment by stuntguy3000** [comment]
@mbax Very true, however I believe it still should be consistent (across commands, and from vanilla  to bukkit).

@turt2live Regarding point 3, a lot of things change from time and time again, so I don't think that is a _valid_ point.

And regarding point 1, I believe it should be strict as clear _can_ be a somewhat powerful command.

**Comment by turt2live** [comment]
Well, from my perspective, the PR wouldn't have been closed if the
reconsideration changed the result.

sent from mobile

**Comment by stuntguy3000** [comment]
The PR was closed, because the ticket was closed. The ticket was closed because I didn't provide an example of what I meant.

**Comment by stuntguy3000** [comment]
"Wesley Wolfe added a comment - Today 4:56 PM
No examples provided of how this is a bug. Being able to use shortened named is a feature, and when combined with features like tab-completion, I don't see any justification that this is a bug."

**Comment by turt2live** [comment]
I fail to see just "no examples" being the only cause in the ticket
closure. Feel free to read.

sent from mobile

**Comment by stuntguy3000** [comment]
Well, I am going by what was said. I don't see any other reasons for the ticket to be closed.

_I do understand the second half of what Wesley said, however I have stated reasons against it_

**Comment by asofold** [comment]
Without judging this PR, having commands like clearing inventories work on prefixes is just asking for trouble. Naturally, an able admin can of course deactivate the command and get a plugin that does not put in "random" players.

The argument "you can tab-complete" should count as an argument for forcing players to write the exact name (case insensitive if you will), rather than to allow short cuts. Short cuts also mean a lot of potential confusion with offline players file present or not present - on bigger servers or just older servers this means .... trouble, potentially.

A difference in handling names from vanilla to Bukkit should make the Bukkit version a bug.

**Comment by literalplus** [comment]
1) Clear will remove items indefinitely and clearing the wrong inventory can lead to lots of trouble from my experience.
Also @asofold, I think @mbax meant that /op etc work with offline players. /clear clearly doesn't.
